File: androidApp/src/main/java/com/example/paradigmaapp/android/viewmodel/EpisodeDetailViewModel.kt
package com.example.paradigmaapp.android.viewmodel

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.paradigmaapp.exception.ApiException
import com.example.paradigmaapp.exception.NoInternetException
import com.example.paradigmaapp.exception.ServerErrorException
import com.example.paradigmaapp.model.Episodio
import com.example.paradigmaapp.model.Programa
import com.example.paradigmaapp.repository.contracts.EpisodioRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel para la pantalla de detalle de un [Episodio].
 * Se encarga de cargar la información completa del episodio especificado por su ID,
 * el cual se recibe a través de [SavedStateHandle] desde los argumentos de navegación.
 * También extrae los programas asociados al episodio a partir de los datos embebidos.
 *
 * @property episodioRepository Repositorio para obtener los datos del episodio.
 * @property savedStateHandle Manejador del estado guardado, utilizado para acceder a los
 * argumentos de navegación (como `episodeId`).
 *
 * @author Mario Alguacil Juárez
 */
class EpisodeDetailViewModel(
    private val episodioRepository: EpisodioRepository,
    private val savedStateHandle: SavedStateHandle
) : ViewModel() {

    /**
     * El ID del episodio cuyos detalles se van a mostrar.
     * Se obtiene de los argumentos de navegación a través de [SavedStateHandle].
     * Si no se encuentra, se establece a -1, indicando un estado inválido.
     */
    val episodeId: Int = savedStateHandle.get<Int>("episodeId") ?: -1

    /**
     * Contiene el estado del objeto [Episodio] actual, con todos sus detalles.
     * Es `null` mientras se carga o si ocurre un error.
     */
    private val _episodio = MutableStateFlow<Episodio?>(null)
    val episodio: StateFlow<Episodio?> = _episodio.asStateFlow()

    /**
     * Contiene el estado de la lista de [Programa]s asociados a este episodio.
     * Se utiliza para mostrar información como los creadores o el programa al que pertenece.
     */
    private val _programasAsociados = MutableStateFlow<List<Programa>>(emptyList())
    val programasAsociados: StateFlow<List<Programa>> = _programasAsociados.asStateFlow()

    /** Indica si los detalles del episodio se están cargando actualmente. */
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    /** Almacena cualquier mensaje de error que ocurra durante la carga de datos. */
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    init {
        // Iniciar la carga de datos solo si se ha proporcionado un ID de episodio válido.
        if (episodeId != -1) {
            loadEpisodeDetails()
        } else {
            _error.value = "ID de episodio no válido."
        }
    }

    /**
     * Carga los detalles completos del episodio utilizando el [episodeId] actual.
     * Obtiene el episodio del [episodioRepository] y extrae los programas asociados
     * de los datos embebidos (`_embedded.wp:term`) del episodio.
     * Actualiza los StateFlows correspondientes ([_episodio], [_programasAsociados], [_isLoading], [_error]).
     */
    fun loadEpisodeDetails() {
        if (episodeId == -1) {
            _error.value = "ID de episodio no válido. No se pueden cargar los detalles."
            return
        }
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null // Limpiar errores previos al iniciar la carga
            try {
                val fetchedEpisodio = episodioRepository.getEpisodio(episodeId)
                _episodio.value = fetchedEpisodio

                if (fetchedEpisodio != null) {
                    // Extraer programas asociados de los datos embebidos (_embedded.wp:term)
                    // que WordPress puede incluir para evitar llamadas de red adicionales.
                    fetchedEpisodio.embedded?.terms?.let { terminosAnidados ->
                        // La API devuelve una lista de listas, una por cada taxonomía. Las aplanamos.
                        val programasEncontrados = terminosAnidados.flatten()
                            .distinctBy { programa -> programa.id }

                        // Filtrar para asegurar que solo se incluyen los programas cuyos IDs están
                        // realmente listados en el campo `radio` del episodio.
                        // Esto proporciona una capa adicional de consistencia.
                        _programasAsociados.value = fetchedEpisodio.programaIds?.let { idsDelEpisodio ->
                            programasEncontrados.filter { programa -> idsDelEpisodio.contains(programa.id) }
                        } ?: programasEncontrados // Fallback a todos los encontrados si programaIds es null
                    } ?: run {
                        // Si no hay datos de términos embebidos, la lista de programas asociados queda vacía.
                        _programasAsociados.value = emptyList()
                    }
                } else {
                    // Si el repositorio devuelve null (ej. por un error 404), establecer un error.
                    if (_error.value == null) { // Solo si no se ha establecido un error más específico.
                        _error.value = "No se pudo encontrar el episodio."
                    }
                }

            } catch (e: NoInternetException) {
                _error.value = e.message ?: "Sin conexión a internet."
            } catch (e: ServerErrorException) {
                _error.value = e.userFriendlyMessage
            } catch (e: ApiException) {
                _error.value = e.message ?: "Error de API al cargar detalles del episodio."
            } catch (e: Exception) {
                _error.value = "Ocurrió un error desconocido al cargar los detalles del episodio."
            } finally {
                _isLoading.value = false
            }
        }
    }
}

File: iosApp/iosApp/viewmodel/EpisodeDetailViewModel.swift
import Foundation
import shared

/**
 * ViewModel para la pantalla que muestra los detalles completos de un episodio.
 *
 * Se encarga de:
 * - Recibir el ID de un episodio.
 * - Cargar todos los datos de ese episodio desde el SDK.
 * - Extraer los programas asociados a partir de los datos embebidos del episodio.
 * - Publicar los datos y los estados de carga/error para la vista de SwiftUI.
 */
@MainActor
class EpisodeDetailViewModel: ObservableObject {

    // MARK: - Propiedades Publicadas
    
    /// El objeto Episodio con todos sus detalles. Es `nil` hasta que se carga.
    @Published var episodio: Episodio? = nil
    
    /// La lista de programas a los que pertenece este episodio.
    @Published var programasAsociados: [Programa] = []
    
    /// Indica si se está realizando la carga inicial de datos para la vista.
    @Published var isLoading = false
    
    /// Contiene un mensaje de error si la carga de datos falla.
    @Published var errorMessage: String? = nil

    // MARK: - Propiedades Privadas

    // Utiliza la instancia única y compartida del SDK.
    private let sdk = AppServices.shared.sdk
    
    /// El ID del episodio que este ViewModel está gestionando.
    private let episodioId: Int

    // MARK: - Inicializador

    /**
     * Inicializa el ViewModel con el ID del episodio a mostrar.
     * - Parameter episodioId: El identificador único del episodio.
     */
    init(episodioId: Int) {
        self.episodioId = episodioId
    }

    // MARK: - Métodos Públicos

    /// Carga los detalles completos del episodio desde el SDK.
    func loadDetails() async {
        // Evita recargar si los datos ya existen.
        guard self.episodio == nil else { return }

        self.isLoading = true
        self.errorMessage = nil

        do {
            // Llama a la función del SDK para obtener el episodio por su ID.
            if let fetchedEpisodio = try await asyncResult(for: sdk.getEpisodio(episodioId: Int32(episodioId))) {
                // Si se encuentra, lo publicamos para la vista.
                self.episodio = fetchedEpisodio
                // Extraemos los programas asociados desde los datos embebidos.
                self.extractAssociatedPrograms(from: fetchedEpisodio)
            } else {
                // Si el SDK devuelve null (ej. error 404), mostramos un mensaje.
                self.errorMessage = "No se pudo encontrar el episodio."
            }
        } catch {
            // Capturamos cualquier otro error de red o de la API.
            self.errorMessage = "Error al cargar los detalles: \(error.localizedDescription)"
        }

        self.isLoading = false
    }

    // MARK: - Métodos Privados

    /**
     * Procesa el objeto Episodio para extraer la lista de programas a los que pertenece.
     * La API de WordPress devuelve esta información en una estructura anidada (`_embedded`).
     * - Parameter from: El objeto `Episodio` del que se extraerán los programas.
     */
    private func extractAssociatedPrograms(from episodio: Episodio) {
        // La propiedad 'terms' es una lista de listas. Primero las aplanamos en una sola.
        let allTerms = episodio.embedded?.terms?.flatMap { $0 } ?? []

        if let programaIds = episodio.programaIds {
            // Creamos un Set de los IDs para una búsqueda más eficiente.
            let idSet = Set(programaIds.map { $0.int32Value })
            // Filtramos los términos para quedarnos solo con los programas correctos.
            self.programasAsociados = allTerms.filter { programa in
                idSet.contains(programa.id)
            }
        } else {
            self.programasAsociados = allTerms
        }
    }
}
